### ISSUE TITLE
Allow to prefer returning array over lists.

### ISSUE THREAD

**Issue by sanelli** [sanelli/Mappa#29]

**Comment by sanelli** [comment]
### Cursor implementation plan

Uploaded plan file: `issue_29_enumerableconcretetype_dcef3fea.plan.md`

---
---
name: Issue 29 EnumerableConcreteType
overview: Add `EnumerableConcreteType` to `MappaSettings` so collection-to-collection mappings targeting sequence-like interfaces (e.g. `IEnumerable<T>`) can allocate `T[]` instead of `List<T>`, with full settings stacking, tests, samples, docs, and two version bumps from `10.1.0-alpha.36`.
todos:
  - id: "1"
    content: 1. Mappa — add EnumerableConcreteTypeSetting enum, extend MappaSettingsAttribute, update Mappa/README.md
    status: completed
  - id: "2"
    content: 2. Bump MappaVersion.targets to alpha.37 and run .\Scripts\RebuildUponVersionChange.ps1
    status: completed
  - id: "3"
    content: "3. Mappa.Generator — wire settings pipeline; extend CollectionToCollectionMapStrategy/detector/builder; remove TODO #29"
    status: completed
  - id: "4"
    content: 4. Mappa.Generator.Tests — add Array behavior and editorconfig/class/method stacking integration tests
    status: completed
  - id: "5"
    content: 5. Bump MappaVersion.targets to alpha.38 and run .\Scripts\RebuildUponVersionChange.ps1
    status: completed
  - id: "6"
    content: 6. Mappa.Samples — add EnumerableConcreteType mappers (List/Array settings × IEnumerable/List return types) and update Mappa.Samples/README.md
    status: completed
  - id: "7"
    content: 7. Mappa.Samples.Tests — assert runtime concrete type is T[] or List<T> per return type and EnumerableConcreteType setup
    status: completed
  - id: "8"
    content: 8. Mappa.Samples.Aot — add runner verifying T[] vs List<T> runtime types and register in Program.cs
    status: completed
  - id: "9"
    content: 9. Documentation — update mappa-attributes.md, tutorial.md, mappa-generator-algorithm.md, features.md, Mappa.Generator/README.md
    status: completed
  - id: "10"
    content: 10. Run .\Scripts\RunTestsAndReportCoverage.ps1 and fix StyleCop/Sonar violations
    status: completed
isProject: false
---

# Issue #29: EnumerableConcreteType (Array vs List)

## Goal

Implement [GitHub issue #29](https://github.com/sanelli/Mappa/issues/29): when a collection mapping targets a sequence-like **interface** (today hidden behind `List<T>`), allow `[MappaSettings]` to choose the concrete buffer type.

Remove the existing TODO in [`ContainerMapStrategyDetector.cs`](Mappa.Generator/Algorithm/StrategyDetectors/ContainerMapStrategyDetector.cs) (line 56).

**Prerequisite:** current version is `10.1.0-alpha.36` in [`MappaVersion.targets`](MappaVersion.targets). Follow the same settings pipeline pattern used by [`IdentityMapDeepCopySetting`](Mappa/IdentityMapDeepCopySetting.cs) (#14) and [`BooleanSetting` collection settings](Mappa/Attributes/MappaSettingsAttribute.cs) (`FastCollections`, `ContainerCapacityConstructors`).

## Issue Requirements (from #29)

| Value | Behavior |
|-------|----------|
| `Undefined` | Inherit from parent scope / `.editorconfig`; does not override stacked settings |
| `List` | **Default** — current behavior (`List<T>` buffer + `Add`) |
| `Array` | Allocate `T[]` (indexer insertion) instead of `List<T>` |

**Required tests (from issue):**
- New `Array` behavior
- Method `[MappaSettings]` overrides class `[MappaSettings]`
- Class `[MappaSettings]` overrides `.editorconfig`

## Scope Boundaries

**In scope**
- New public enum + `MappaSettings` property
- Full settings stack: `.editorconfig` → global options → class → method
- [`CollectionToCollectionMapStrategy`](Mappa.Generator/Models/Strategies/CollectionToCollectionMapStrategy.cs) / [`CollectionToCollectionMapStrategyBuilder`](Mappa.Generator/Builders/Strategies/CollectionToCollectionMapStrategyBuilder.cs) changes
- Generator integration tests, runtime sample + unit tests, AOT runner, documentation
- Two alpha version bumps + rebuilds

**Targets affected when `Array` is effective** (interface targets that currently use the `List<T>` branch at lines 347–368):

| Target interface | `Array` behavior |
|------------------|------------------|
| `IEnumerable<T>` | `T[]` buffer, returned as `IEnumerable<T>` |
| `IList<T>` | `T[]` buffer (`T[]` implements `IList<T>`) |
| `IReadOnlyList<T>` | `T[]` buffer |
| `ICollection<T>` | `T[]` buffer |
| `IReadOnlyCollection<T>` | `T[]` buffer |

**Out of scope (unchanged behavior)**

| Case | Reason |
|------|--------|
| Concrete `List<T>` target | User explicitly requested `List<T>` |
| `T[]`, `Span<T>`, `Memory<T>`, stacks, queues, sets, dictionaries | Already have dedicated allocation paths |
| `ReadOnlyCollection<T>`, immutable/frozen types in the List branch | Different concrete semantics; post-loop wrappers expect a growable buffer today |
| Dictionary strategy | Separate detector branch |
| New diagnostics | Issue does not request any |

```mermaid
flowchart TD
    detect[ContainerMapStrategyDetector]
    coll[CollectionToCollectionMapStrategy]
    resolve[Resolve effective EnumerableConcreteType]
    targetKind{Target kind}
    listBuf[List buffer + Add]
    arrayBuf[T array + indexer]
    build[CollectionToCollectionMapStrategyBuilder]

    detect --> coll
    coll --> resolve
    resolve --> targetKind
    targetKind -->|interface sequence + Array| arrayBuf
    targetKind -->|otherwise| listBuf
    arrayBuf --> build
    listBuf --> build
```

## Settings Design

### New enum: `EnumerableConcreteTypeSetting`

Add [`Mappa/EnumerableConcreteTypeSetting.cs`](Mappa/EnumerableConcreteTypeSetting.cs) (mirror [`IdentityMapDeepCopySetting.cs`](Mappa/IdentityMapDeepCopySetting.cs)):

```csharp
public enum EnumerableConcreteTypeSetting
{
    Undefined,
    List,
    Array,
}
```

### New `MappaSettings` property

Add to [`Mappa/Attributes/MappaSettingsAttribute.cs`](Mappa/Attributes/MappaSettingsAttribute.cs):

- **Property:** `EnumerableConcreteType` (type `EnumerableConcreteTypeSetting`, default `Undefined`)
- **Effective default:** `List` when unset everywhere
- **`.editorconfig` key:** `mappa.enumerableconcretetype`
- **Allowed values:** `List`, `Array`, `Undefined`

Document in [`Mappa/README.md`](Mappa/README.md) and [`Documentation/mappa-attributes.md`](Documentation/mappa-attributes.md).

## Generator Architecture

### 1. Settings pipeline (Generator)

Wire through the same touch points as `IdentityMapDeepCopy` / `FastCollections`:

| File | Change |
|------|--------|
| [`Mappa.Generator/Models/IMappaUserSettings.cs`](Mappa.Generator/Models/IMappaUserSettings.cs) | Add `EnumerableConcreteTypeSetting EnumerableConcreteType { get; }` |
| [`Mappa.Generator/Models/MappaUserSettings.cs`](Mappa.Generator/Models/MappaUserSettings.cs) | `StackSetting<EnumerableConcreteTypeSetting>`, ctor param, property, `Apply(...)` |
| [`Mappa.Generator/Models/MappaGlobalOptions.cs`](Mappa.Generator/Models/MappaGlobalOptions.cs) | Parse `mappa.enumerableconcretetype`; default `List`; XML doc entry |
| [`Mappa.Generator/Extensions/AttributeDataExtensions.cs`](Mappa.Generator/Extensions/AttributeDataExtensions.cs) | Read `nameof(MappaSettingsAttribute.EnumerableConcreteType)` |

Add helper:

```csharp
static EnumerableConcreteTypeSetting GetEffective(EnumerableConcreteTypeSetting value)
    => value is EnumerableConcreteTypeSetting.Undefined
        ? EnumerableConcreteTypeSetting.List
        : value;
```

### 2. Strategy model

Update [`CollectionToCollectionMapStrategy.cs`](Mappa.Generator/Models/Strategies/CollectionToCollectionMapStrategy.cs):

- Add `EnumerableConcreteTypeSetting EnumerableConcreteType` constructor parameter + property
- Pass effective setting from [`ContainerMapStrategyDetector.cs`](Mappa.Generator/Algorithm/StrategyDetectors/ContainerMapStrategyDetector.cs) (replace TODO `#29`)

### 3. Code generation

Update [`CollectionToCollectionMapStrategyBuilder.cs`](Mappa.Generator/Builders/Strategies/CollectionToCollectionMapStrategyBuilder.cs):

**`AppendTargetVariable`** — add `EnumerableConcreteTypeSetting enumerableConcreteType` parameter (read from strategy).

Before the existing List branch (lines 347–368), when **all** of the following hold:
1. Effective setting is `Array`
2. Target `TypeKind` is `Interface`
3. Target is a sequence interface (`IsIEnumerable()`, `IsIList()`, `IsIReadOnlyList()`, `IsICollection()`, or `IsIReadOnlyCollection()`)

…use the **existing array allocation path** (lines 220–245):
- `InsertionMethod.Indexer`
- `T[] __tmp = new T[capacity];`
- Counter variable when source lacks indexer (same as `IEnumerable<T>` → `T[]` today)

Otherwise keep current `List<T>` + `Add` behavior.

**`FastCollections` interaction:** once the buffer is `T[]`, the existing fast-collection paths (span/`for` for array/list sources, `AsSpan()` target when source is array/list) apply naturally — no separate work required beyond choosing the array buffer.

**`AppendPostLoopCode`:** no changes needed for the five sequence interfaces (array is returned directly). Immutable/set targets in the List branch remain List-backed.

Extract a small private static helper (e.g. `ShouldUseArrayForEnumerableInterfaceTarget`) to keep `AppendTargetVariable` readable and testable by inspection.

## Tests

Add [`Mappa.Generator.Tests/CollectionToCollectionMapStrategyIntegrationTests.EnumerableConcreteType.cs`](Mappa.Generator.Tests/CollectionToCollectionMapStrategyIntegrationTests.EnumerableConcreteType.cs) (partial class).

### Core behavior

1. **Default List (regression)** — existing [`CanMapFromIEnumerableToIEnumerable`](Mappa.Generator.Tests/CollectionToCollectionMapStrategyIntegrationTests.cs) remains valid; no change expected
2. **Array on method** — `[MappaSettings(EnumerableConcreteType = Array)]` on `IEnumerable<int> → IEnumerable<string>`; assert `string[]` local declaration + indexer assignment (not `List<string>.Add`)
3. **Array on class** — class-level attribute, method unannotated
4. **Concrete `List<T>` unchanged** — `[MappaSettings(EnumerableConcreteType = Array)]` on `IEnumerable<int> → List<string>`; still asserts `List<string>` creation
5. **Optional extra coverage** — one `ICollection<T>` or `IReadOnlyList<T>` interface target with `Array` to confirm the shared branch

### Settings stacking (required by issue)

Use `IEnumerable<T> → IEnumerable<T>` and assert generated syntax (not just diagnostics):

| Test | `.editorconfig` | Class `[MappaSettings]` | Method `[MappaSettings]` | Expected |
|------|-----------------|-------------------------|--------------------------|----------|
| **Editorconfig only** | `mappa.enumerableconcretetype = Array` | *(none)* | *(none)* | `string[]` buffer |
| **Class overrides editorconfig** | `Array` | `List` | *(none)* | `List<string>` buffer |
| **Method overrides class and editorconfig** | `Array` | `List` | `Array` | `string[]` buffer |

Suggested names (follow [`IdentityStrategyIntegrationTests`](Mappa.Generator.Tests/IdentityStrategyIntegrationTests.cs) stacking pattern):

- `CanMapFromIEnumerableToIEnumerableWhenEnumerableConcreteTypeIsSetInEditorConfig`
- `CanMapFromIEnumerableToIEnumerableWhenClassEnumerableConcreteTypeOverridesEditorConfig`
- `CanMapFromIEnumerableToIEnumerableWhenMethodEnumerableConcreteTypeOverridesClassAndEditorConfig`

Follow `HasSyntaxNodesCount(X)` / `HasNextSyntaxNode` chaining rules from workflow.

**No `Mappa.Tests` changes** — enum settings are covered by generator integration tests (same as `IdentityMapDeepCopy`).

## Samples

Add dedicated mappers in [`Mappa.Samples/EnumerableConcreteTypeMapper.cs`](Mappa.Samples/EnumerableConcreteTypeMapper.cs) so **return type** and **setting** can be exercised independently (do not rely on [`CollectionToCollectionMapper.cs`](Mappa.Samples/CollectionToCollectionMapper.cs) alone — it has no `EnumerableConcreteType` settings and existing tests only assert element equivalence, not concrete type).

| Mapper class | `[MappaSettings]` | Map method return type | Expected runtime type |
|--------------|-------------------|------------------------|------------------------|
| `EnumerableConcreteTypeListMapper` | *(none — default `List`)* | `IEnumerable<int>` | `List<int>` |
| `EnumerableConcreteTypeArrayMapper` | `EnumerableConcreteType = Array` | `IEnumerable<int>` | `int[]` |
| `EnumerableConcreteTypeExplicitListMapper` | `EnumerableConcreteType = Array` | `List<int>` | `List<int>` (concrete target wins) |
| `EnumerableConcreteTypeArrayInterfaceMapper` | `EnumerableConcreteType = Array` | `IList<int>` or `ICollection<int>` | `int[]` |

Use separate partial mapper classes (same pattern as [`IdentityMapDeepCopyMapper.cs`](Mappa.Samples/IdentityMapDeepCopyMapper.cs)) so each class carries a single effective setting.

Update [`Mappa.Samples/README.md`](Mappa.Samples/README.md).

### [`Mappa.Samples.Tests`](Mappa.Samples.Tests) — required runtime concrete-type assertions

Add [`EnumerableConcreteTypeMapperUnitTest.cs`](Mappa.Samples.Tests/EnumerableConcreteTypeMapperUnitTest.cs).

**Every test must assert the runtime concrete type of the mapped result**, not only element equivalence. Use `Assert.IsType<T>(result)` on the object returned by the map method (the compile-time return type may be an interface, but the runtime type must match the setting).

| Test | Mapper | Assertion |
|------|--------|-----------|
| Default `List` on `IEnumerable<T>` | `EnumerableConcreteTypeListMapper` | `Assert.IsType<List<int>>(result)` |
| `Array` on `IEnumerable<T>` | `EnumerableConcreteTypeArrayMapper` | `Assert.IsType<int[]>(result)` |
| `Array` setting but `List<T>` return | `EnumerableConcreteTypeExplicitListMapper` | `Assert.IsType<List<int>>(result)` and `Assert.IsNotType<int[]>(result)` |
| `Array` on `IList<T>` / `ICollection<T>` | `EnumerableConcreteTypeArrayInterfaceMapper` | `Assert.IsType<int[]>(result)` |

Also assert mapped **content** (e.g. `Assert.Equal` / `BeEquivalentTo`) so type checks are not vacuous.

**Do not** use `result.ToList()` before the type assertion on `IEnumerable<T>` returns — that would mask whether the mapper already returned a `List<T>` or `T[]`.

Example pattern:

```csharp
var result = arrayMapper.Map(input);
var arrayResult = Assert.IsType<int[]>(result);
Assert.Equal([0, 2], arrayResult);
```

```csharp
var result = listMapper.Map(input);
var listResult = Assert.IsType<List<int>>(result);
Assert.Equal([0, 2], listResult);
```

### [`Mappa.Samples.Aot`](Mappa.Samples.Aot)

Add [`Runners/EnumerableConcreteTypeMapperRunner.cs`](Mappa.Samples.Aot/Runners/EnumerableConcreteTypeMapperRunner.cs) and register in [`Program.cs`](Mappa.Samples.Aot/Program.cs).

The AOT runner must exercise the same scenarios and **report the runtime type** of each mapped result (e.g. via `result.GetType().Name` in the runner report, or `Debug.Assert` / throw if type mismatches), mirroring the unit-test matrix above.

## Documentation

| File | Update |
|------|--------|
| [`Documentation/mappa-attributes.md`](Documentation/mappa-attributes.md) | Property list + editorconfig table row |
| [`Documentation/tutorial.md`](Documentation/tutorial.md) | Short section with `IEnumerable<T>` → array example |
| [`Documentation/mappa-generator-algorithm.md`](Documentation/mappa-generator-algorithm.md) | Collection strategy notes (interface buffer selection) |
| [`Documentation/features.md`](Documentation/features.md) | New row under MappaSettings table |
| [`Mappa.Generator/README.md`](Mappa.Generator/README.md) | Collection strategy section |

No changes to [`Documentation/error-codes.md`](Documentation/error-codes.md) unless a diagnostic is added.

## Versioning

| Step | Bump | When |
|------|------|------|
| After **Mappa** | `10.1.0-alpha.37` | Before Generator work |
| After **Mappa.Generator** (feature complete, tests pass) | `10.1.0-alpha.38` | Before Samples |

Run `.\Scripts\RebuildUponVersionChange.ps1` after each bump.

## Implementation Order (workflow)

### 1. Mappa — add `EnumerableConcreteTypeSetting`, extend `MappaSettingsAttribute`, update `Mappa/README.md`

### 2. Bump `MappaVersion.targets` to `alpha.37` and run `.\Scripts\RebuildUponVersionChange.ps1`

### 3. Mappa.Generator — settings pipeline; extend `CollectionToCollectionMapStrategy` / detector / builder; remove TODO `#29`

### 4. Mappa.Generator.Tests — integration tests listed above (core + three stacking tests)

### 5. Bump `MappaVersion.targets` to `alpha.38` and run `.\Scripts\RebuildUponVersionChange.ps1`

### 6. Mappa.Samples — `EnumerableConcreteTypeMapper` + README

### 7. Mappa.Samples.Tests — assert runtime concrete type is `T[]` or `List<T>` per return type and `EnumerableConcreteType` setup

### 8. Mappa.Samples.Aot — runner + `Program.cs` registration

### 9. Documentation — attributes, tutorial, algorithm, features, generator README

### 10. Run `.\Scripts\RunTestsAndReportCoverage.ps1` and fix StyleCop/Sonar violations

## Key Files Reference

Current List allocation (change point):

```347:368:Mappa.Generator/Builders/Strategies/CollectionToCollectionMapStrategyBuilder.cs
        else if (targetTypeSymbol.IsIEnumerable()
            || targetTypeSymbol.IsList(context.Compilation)
            // ...
        {
            insertionMethod = InsertionMethod.Add;
            TryGetLengthExpressionFromProperty(source, sourceTypeSymbol, context.Compilation, out var capacity);
            stringBuilder.AppendLine($"global::System.Collections.Generic.List<...> {targetVariableName} = new ...({capacity});");
        }
```

Existing array allocation to reuse:

```220:245:Mappa.Generator/Builders/Strategies/CollectionToCollectionMapStrategyBuilder.cs
        else if (targetTypeSymbol.IsArray()
                  || targetTypeSymbol.IsSpan(context.Compilation)
                  // ...
        {
            insertionMethod = InsertionMethod.Indexer;
            var capacity = GetLengthExpression(source, sourceTypeSymbol, context.Compilation);
            stringBuilder.AppendLine($"{elementType}[] {targetVariableName} = new {elementType}[{capacity}];");
            if (!HasIndexer(context, sourceTypeSymbol))
            {
                counterVariableName = context.NextTemporary();
                stringBuilder.AppendLine($"int {counterVariableName} = 0;");
            }
        }
```

TODO to remove:

```55:66:Mappa.Generator/Algorithm/StrategyDetectors/ContainerMapStrategyDetector.cs
        // 02. Collection -> Collection strategy.
        // TODO [#29] Allow to prefer returning array over lists for interfaces.
        else if (this.CanMapCollectionToCollection(out var elementStrategy))
        {
            mapStrategy = new CollectionToCollectionMapStrategy(
                // ... add EnumerableConcreteType from context.MappaUserSettings
```

## Risks and Mitigations

| Risk | Mitigation |
|------|------------|
| Double enumeration on generic `IEnumerable<T>` sources (capacity via `Count()`) | Same as existing `IEnumerable` → `T[]` path; documented in algorithm notes (related TODO #108) |
| Breaking change | Default remains `List`; all existing tests must pass unchanged |
| `FastCollections` + `Array` combo | Array buffer unlocks existing fast paths automatically |
| Immutable/set types accidentally switched to array | Restrict `Array` to `TypeKind.Interface` sequence interfaces only |

**Comment by sanelli** [comment]
### Cursor implementation plan

Uploaded plan file: `issue_29_enumerableconcretetype_dcef3fea.plan.md`

---
---
name: Issue 29 EnumerableConcreteType
overview: Add `EnumerableConcreteType` to `MappaSettings` so collection-to-collection mappings targeting sequence-like interfaces (e.g. `IEnumerable<T>`) can allocate `T[]` instead of `List<T>`, with full settings stacking, tests, samples, docs, and two version bumps from `10.1.0-alpha.36`.
todos:
  - id: "1"
    content: 1. Mappa — add EnumerableConcreteTypeSetting enum, extend MappaSettingsAttribute, update Mappa/README.md
    status: completed
  - id: "2"
    content: 2. Bump MappaVersion.targets to alpha.37 and run .\Scripts\RebuildUponVersionChange.ps1
    status: completed
  - id: "3"
    content: "3. Mappa.Generator — wire settings pipeline; extend CollectionToCollectionMapStrategy/detector/builder; remove TODO #29"
    status: completed
  - id: "4"
    content: 4. Mappa.Generator.Tests — add Array behavior and editorconfig/class/method stacking integration tests
    status: completed
  - id: "5"
    content: 5. Bump MappaVersion.targets to alpha.38 and run .\Scripts\RebuildUponVersionChange.ps1
    status: completed
  - id: "6"
    content: 6. Mappa.Samples — add EnumerableConcreteType mappers (List/Array settings × IEnumerable/List return types) and update Mappa.Samples/README.md
    status: completed
  - id: "7"
    content: 7. Mappa.Samples.Tests — assert runtime concrete type is T[] or List<T> per return type and EnumerableConcreteType setup
    status: completed
  - id: "8"
    content: 8. Mappa.Samples.Aot — add runner verifying T[] vs List<T> runtime types and register in Program.cs
    status: completed
  - id: "9"
    content: 9. Documentation — update mappa-attributes.md, tutorial.md, mappa-generator-algorithm.md, features.md, Mappa.Generator/README.md
    status: completed
  - id: "10"
    content: 10. Run .\Scripts\RunTestsAndReportCoverage.ps1 and fix StyleCop/Sonar violations
    status: completed
isProject: false
---

# Issue #29: EnumerableConcreteType (Array vs List)

## Goal

Implement [GitHub issue #29](https://github.com/sanelli/Mappa/issues/29): when a collection mapping targets a sequence-like **interface** (today hidden behind `List<T>`), allow `[MappaSettings]` to choose the concrete buffer type.

Remove the existing TODO in [`ContainerMapStrategyDetector.cs`](Mappa.Generator/Algorithm/StrategyDetectors/ContainerMapStrategyDetector.cs) (line 56).

**Prerequisite:** current version is `10.1.0-alpha.36` in [`MappaVersion.targets`](MappaVersion.targets). Follow the same settings pipeline pattern used by [`IdentityMapDeepCopySetting`](Mappa/IdentityMapDeepCopySetting.cs) (#14) and [`BooleanSetting` collection settings](Mappa/Attributes/MappaSettingsAttribute.cs) (`FastCollections`, `ContainerCapacityConstructors`).

## Issue Requirements (from #29)

| Value | Behavior |
|-------|----------|
| `Undefined` | Inherit from parent scope / `.editorconfig`; does not override stacked settings |
| `List` | **Default** — current behavior (`List<T>` buffer + `Add`) |
| `Array` | Allocate `T[]` (indexer insertion) instead of `List<T>` |

**Required tests (from issue):**
- New `Array` behavior
- Method `[MappaSettings]` overrides class `[MappaSettings]`
- Class `[MappaSettings]` overrides `.editorconfig`

## Scope Boundaries

**In scope**
- New public enum + `MappaSettings` property
- Full settings stack: `.editorconfig` → global options → class → method
- [`CollectionToCollectionMapStrategy`](Mappa.Generator/Models/Strategies/CollectionToCollectionMapStrategy.cs) / [`CollectionToCollectionMapStrategyBuilder`](Mappa.Generator/Builders/Strategies/CollectionToCollectionMapStrategyBuilder.cs) changes
- Generator integration tests, runtime sample + unit tests, AOT runner, documentation
- Two alpha version bumps + rebuilds

**Targets affected when `Array` is effective** (interface targets that currently use the `List<T>` branch at lines 347–368):

| Target interface | `Array` behavior |
|------------------|------------------|
| `IEnumerable<T>` | `T[]` buffer, returned as `IEnumerable<T>` |
| `IList<T>` | `T[]` buffer (`T[]` implements `IList<T>`) |
| `IReadOnlyList<T>` | `T[]` buffer |
| `ICollection<T>` | `T[]` buffer |
| `IReadOnlyCollection<T>` | `T[]` buffer |

**Out of scope (unchanged behavior)**

| Case | Reason |
|------|--------|
| Concrete `List<T>` target | User explicitly requested `List<T>` |
| `T[]`, `Span<T>`, `Memory<T>`, stacks, queues, sets, dictionaries | Already have dedicated allocation paths |
| `ReadOnlyCollection<T>`, immutable/frozen types in the List branch | Different concrete semantics; post-loop wrappers expect a growable buffer today |
| Dictionary strategy | Separate detector branch |
| New diagnostics | Issue does not request any |

```mermaid
flowchart TD
    detect[ContainerMapStrategyDetector]
    coll[CollectionToCollectionMapStrategy]
    resolve[Resolve effective EnumerableConcreteType]
    targetKind{Target kind}
    listBuf[List buffer + Add]
    arrayBuf[T array + indexer]
    build[CollectionToCollectionMapStrategyBuilder]

    detect --> coll
    coll --> resolve
    resolve --> targetKind
    targetKind -->|interface sequence + Array| arrayBuf
    targetKind -->|otherwise| listBuf
    arrayBuf --> build
    listBuf --> build
```

## Settings Design

### New enum: `EnumerableConcreteTypeSetting`

Add [`Mappa/EnumerableConcreteTypeSetting.cs`](Mappa/EnumerableConcreteTypeSetting.cs) (mirror [`IdentityMapDeepCopySetting.cs`](Mappa/IdentityMapDeepCopySetting.cs)):

```csharp
public enum EnumerableConcreteTypeSetting
{
    Undefined,
    List,
    Array,
}
```

### New `MappaSettings` property

Add to [`Mappa/Attributes/MappaSettingsAttribute.cs`](Mappa/Attributes/MappaSettingsAttribute.cs):

- **Property:** `EnumerableConcreteType` (type `EnumerableConcreteTypeSetting`, default `Undefined`)
- **Effective default:** `List` when unset everywhere
- **`.editorconfig` key:** `mappa.enumerableconcretetype`
- **Allowed values:** `List`, `Array`, `Undefined`

Document in [`Mappa/README.md`](Mappa/README.md) and [`Documentation/mappa-attributes.md`](Documentation/mappa-attributes.md).

## Generator Architecture

### 1. Settings pipeline (Generator)

Wire through the same touch points as `IdentityMapDeepCopy` / `FastCollections`:

| File | Change |
|------|--------|
| [`Mappa.Generator/Models/IMappaUserSettings.cs`](Mappa.Generator/Models/IMappaUserSettings.cs) | Add `EnumerableConcreteTypeSetting EnumerableConcreteType { get; }` |
| [`Mappa.Generator/Models/MappaUserSettings.cs`](Mappa.Generator/Models/MappaUserSettings.cs) | `StackSetting<EnumerableConcreteTypeSetting>`, ctor param, property, `Apply(...)` |
| [`Mappa.Generator/Models/MappaGlobalOptions.cs`](Mappa.Generator/Models/MappaGlobalOptions.cs) | Parse `mappa.enumerableconcretetype`; default `List`; XML doc entry |
| [`Mappa.Generator/Extensions/AttributeDataExtensions.cs`](Mappa.Generator/Extensions/AttributeDataExtensions.cs) | Read `nameof(MappaSettingsAttribute.EnumerableConcreteType)` |

Add helper:

```csharp
static EnumerableConcreteTypeSetting GetEffective(EnumerableConcreteTypeSetting value)
    => value is EnumerableConcreteTypeSetting.Undefined
        ? EnumerableConcreteTypeSetting.List
        : value;
```

### 2. Strategy model

Update [`CollectionToCollectionMapStrategy.cs`](Mappa.Generator/Models/Strategies/CollectionToCollectionMapStrategy.cs):

- Add `EnumerableConcreteTypeSetting EnumerableConcreteType` constructor parameter + property
- Pass effective setting from [`ContainerMapStrategyDetector.cs`](Mappa.Generator/Algorithm/StrategyDetectors/ContainerMapStrategyDetector.cs) (replace TODO `#29`)

### 3. Code generation

Update [`CollectionToCollectionMapStrategyBuilder.cs`](Mappa.Generator/Builders/Strategies/CollectionToCollectionMapStrategyBuilder.cs):

**`AppendTargetVariable`** — add `EnumerableConcreteTypeSetting enumerableConcreteType` parameter (read from strategy).

Before the existing List branch (lines 347–368), when **all** of the following hold:
1. Effective setting is `Array`
2. Target `TypeKind` is `Interface`
3. Target is a sequence interface (`IsIEnumerable()`, `IsIList()`, `IsIReadOnlyList()`, `IsICollection()`, or `IsIReadOnlyCollection()`)

…use the **existing array allocation path** (lines 220–245):
- `InsertionMethod.Indexer`
- `T[] __tmp = new T[capacity];`
- Counter variable when source lacks indexer (same as `IEnumerable<T>` → `T[]` today)

Otherwise keep current `List<T>` + `Add` behavior.

**`FastCollections` interaction:** once the buffer is `T[]`, the existing fast-collection paths (span/`for` for array/list sources, `AsSpan()` target when source is array/list) apply naturally — no separate work required beyond choosing the array buffer.

**`AppendPostLoopCode`:** no changes needed for the five sequence interfaces (array is returned directly). Immutable/set targets in the List branch remain List-backed.

Extract a small private static helper (e.g. `ShouldUseArrayForEnumerableInterfaceTarget`) to keep `AppendTargetVariable` readable and testable by inspection.

## Tests

Add [`Mappa.Generator.Tests/CollectionToCollectionMapStrategyIntegrationTests.EnumerableConcreteType.cs`](Mappa.Generator.Tests/CollectionToCollectionMapStrategyIntegrationTests.EnumerableConcreteType.cs) (partial class).

### Core behavior

1. **Default List (regression)** — existing [`CanMapFromIEnumerableToIEnumerable`](Mappa.Generator.Tests/CollectionToCollectionMapStrategyIntegrationTests.cs) remains valid; no change expected
2. **Array on method** — `[MappaSettings(EnumerableConcreteType = Array)]` on `IEnumerable<int> → IEnumerable<string>`; assert `string[]` local declaration + indexer assignment (not `List<string>.Add`)
3. **Array on class** — class-level attribute, method unannotated
4. **Concrete `List<T>` unchanged** — `[MappaSettings(EnumerableConcreteType = Array)]` on `IEnumerable<int> → List<string>`; still asserts `List<string>` creation
5. **Optional extra coverage** — one `ICollection<T>` or `IReadOnlyList<T>` interface target with `Array` to confirm the shared branch

### Settings stacking (required by issue)

Use `IEnumerable<T> → IEnumerable<T>` and assert generated syntax (not just diagnostics):

| Test | `.editorconfig` | Class `[MappaSettings]` | Method `[MappaSettings]` | Expected |
|------|-----------------|-------------------------|--------------------------|----------|
| **Editorconfig only** | `mappa.enumerableconcretetype = Array` | *(none)* | *(none)* | `string[]` buffer |
| **Class overrides editorconfig** | `Array` | `List` | *(none)* | `List<string>` buffer |
| **Method overrides class and editorconfig** | `Array` | `List` | `Array` | `string[]` buffer |

Suggested names (follow [`IdentityStrategyIntegrationTests`](Mappa.Generator.Tests/IdentityStrategyIntegrationTests.cs) stacking pattern):

- `CanMapFromIEnumerableToIEnumerableWhenEnumerableConcreteTypeIsSetInEditorConfig`
- `CanMapFromIEnumerableToIEnumerableWhenClassEnumerableConcreteTypeOverridesEditorConfig`
- `CanMapFromIEnumerableToIEnumerableWhenMethodEnumerableConcreteTypeOverridesClassAndEditorConfig`

Follow `HasSyntaxNodesCount(X)` / `HasNextSyntaxNode` chaining rules from workflow.

**No `Mappa.Tests` changes** — enum settings are covered by generator integration tests (same as `IdentityMapDeepCopy`).

## Samples

Add dedicated mappers in [`Mappa.Samples/EnumerableConcreteTypeMapper.cs`](Mappa.Samples/EnumerableConcreteTypeMapper.cs) so **return type** and **setting** can be exercised independently (do not rely on [`CollectionToCollectionMapper.cs`](Mappa.Samples/CollectionToCollectionMapper.cs) alone — it has no `EnumerableConcreteType` settings and existing tests only assert element equivalence, not concrete type).

| Mapper class | `[MappaSettings]` | Map method return type | Expected runtime type |
|--------------|-------------------|------------------------|------------------------|
| `EnumerableConcreteTypeListMapper` | *(none — default `List`)* | `IEnumerable<int>` | `List<int>` |
| `EnumerableConcreteTypeArrayMapper` | `EnumerableConcreteType = Array` | `IEnumerable<int>` | `int[]` |
| `EnumerableConcreteTypeExplicitListMapper` | `EnumerableConcreteType = Array` | `List<int>` | `List<int>` (concrete target wins) |
| `EnumerableConcreteTypeArrayInterfaceMapper` | `EnumerableConcreteType = Array` | `IList<int>` or `ICollection<int>` | `int[]` |

Use separate partial mapper classes (same pattern as [`IdentityMapDeepCopyMapper.cs`](Mappa.Samples/IdentityMapDeepCopyMapper.cs)) so each class carries a single effective setting.

Update [`Mappa.Samples/README.md`](Mappa.Samples/README.md).

### [`Mappa.Samples.Tests`](Mappa.Samples.Tests) — required runtime concrete-type assertions

Add [`EnumerableConcreteTypeMapperUnitTest.cs`](Mappa.Samples.Tests/EnumerableConcreteTypeMapperUnitTest.cs).

**Every test must assert the runtime concrete type of the mapped result**, not only element equivalence. Use `Assert.IsType<T>(result)` on the object returned by the map method (the compile-time return type may be an interface, but the runtime type must match the setting).

| Test | Mapper | Assertion |
|------|--------|-----------|
| Default `List` on `IEnumerable<T>` | `EnumerableConcreteTypeListMapper` | `Assert.IsType<List<int>>(result)` |
| `Array` on `IEnumerable<T>` | `EnumerableConcreteTypeArrayMapper` | `Assert.IsType<int[]>(result)` |
| `Array` setting but `List<T>` return | `EnumerableConcreteTypeExplicitListMapper` | `Assert.IsType<List<int>>(result)` and `Assert.IsNotType<int[]>(result)` |
| `Array` on `IList<T>` / `ICollection<T>` | `EnumerableConcreteTypeArrayInterfaceMapper` | `Assert.IsType<int[]>(result)` |

Also assert mapped **content** (e.g. `Assert.Equal` / `BeEquivalentTo`) so type checks are not vacuous.

**Do not** use `result.ToList()` before the type assertion on `IEnumerable<T>` returns — that would mask whether the mapper already returned a `List<T>` or `T[]`.

Example pattern:

```csharp
var result = arrayMapper.Map(input);
var arrayResult = Assert.IsType<int[]>(result);
Assert.Equal([0, 2], arrayResult);
```

```csharp
var result = listMapper.Map(input);
var listResult = Assert.IsType<List<int>>(result);
Assert.Equal([0, 2], listResult);
```

### [`Mappa.Samples.Aot`](Mappa.Samples.Aot)

Add [`Runners/EnumerableConcreteTypeMapperRunner.cs`](Mappa.Samples.Aot/Runners/EnumerableConcreteTypeMapperRunner.cs) and register in [`Program.cs`](Mappa.Samples.Aot/Program.cs).

The AOT runner must exercise the same scenarios and **report the runtime type** of each mapped result (e.g. via `result.GetType().Name` in the runner report, or `Debug.Assert` / throw if type mismatches), mirroring the unit-test matrix above.

## Documentation

| File | Update |
|------|--------|
| [`Documentation/mappa-attributes.md`](Documentation/mappa-attributes.md) | Property list + editorconfig table row |
| [`Documentation/tutorial.md`](Documentation/tutorial.md) | Short section with `IEnumerable<T>` → array example |
| [`Documentation/mappa-generator-algorithm.md`](Documentation/mappa-generator-algorithm.md) | Collection strategy notes (interface buffer selection) |
| [`Documentation/features.md`](Documentation/features.md) | New row under MappaSettings table |
| [`Mappa.Generator/README.md`](Mappa.Generator/README.md) | Collection strategy section |

No changes to [`Documentation/error-codes.md`](Documentation/error-codes.md) unless a diagnostic is added.

## Versioning

| Step | Bump | When |
|------|------|------|
| After **Mappa** | `10.1.0-alpha.37` | Before Generator work |
| After **Mappa.Generator** (feature complete, tests pass) | `10.1.0-alpha.38` | Before Samples |

Run `.\Scripts\RebuildUponVersionChange.ps1` after each bump.

## Implementation Order (workflow)

### 1. Mappa — add `EnumerableConcreteTypeSetting`, extend `MappaSettingsAttribute`, update `Mappa/README.md`

### 2. Bump `MappaVersion.targets` to `alpha.37` and run `.\Scripts\RebuildUponVersionChange.ps1`

### 3. Mappa.Generator — settings pipeline; extend `CollectionToCollectionMapStrategy` / detector / builder; remove TODO `#29`

### 4. Mappa.Generator.Tests — integration tests listed above (core + three stacking tests)

### 5. Bump `MappaVersion.targets` to `alpha.38` and run `.\Scripts\RebuildUponVersionChange.ps1`

### 6. Mappa.Samples — `EnumerableConcreteTypeMapper` + README

### 7. Mappa.Samples.Tests — assert runtime concrete type is `T[]` or `List<T>` per return type and `EnumerableConcreteType` setup

### 8. Mappa.Samples.Aot — runner + `Program.cs` registration

### 9. Documentation — attributes, tutorial, algorithm, features, generator README

### 10. Run `.\Scripts\RunTestsAndReportCoverage.ps1` and fix StyleCop/Sonar violations

## Key Files Reference

Current List allocation (change point):

```347:368:Mappa.Generator/Builders/Strategies/CollectionToCollectionMapStrategyBuilder.cs
        else if (targetTypeSymbol.IsIEnumerable()
            || targetTypeSymbol.IsList(context.Compilation)
            // ...
        {
            insertionMethod = InsertionMethod.Add;
            TryGetLengthExpressionFromProperty(source, sourceTypeSymbol, context.Compilation, out var capacity);
            stringBuilder.AppendLine($"global::System.Collections.Generic.List<...> {targetVariableName} = new ...({capacity});");
        }
```

Existing array allocation to reuse:

```220:245:Mappa.Generator/Builders/Strategies/CollectionToCollectionMapStrategyBuilder.cs
        else if (targetTypeSymbol.IsArray()
                  || targetTypeSymbol.IsSpan(context.Compilation)
                  // ...
        {
            insertionMethod = InsertionMethod.Indexer;
            var capacity = GetLengthExpression(source, sourceTypeSymbol, context.Compilation);
            stringBuilder.AppendLine($"{elementType}[] {targetVariableName} = new {elementType}[{capacity}];");
            if (!HasIndexer(context, sourceTypeSymbol))
            {
                counterVariableName = context.NextTemporary();
                stringBuilder.AppendLine($"int {counterVariableName} = 0;");
            }
        }
```

TODO to remove:

```55:66:Mappa.Generator/Algorithm/StrategyDetectors/ContainerMapStrategyDetector.cs
        // 02. Collection -> Collection strategy.
        // TODO [#29] Allow to prefer returning array over lists for interfaces.
        else if (this.CanMapCollectionToCollection(out var elementStrategy))
        {
            mapStrategy = new CollectionToCollectionMapStrategy(
                // ... add EnumerableConcreteType from context.MappaUserSettings
```

## Risks and Mitigations

| Risk | Mitigation |
|------|------------|
| Double enumeration on generic `IEnumerable<T>` sources (capacity via `Count()`) | Same as existing `IEnumerable` → `T[]` path; documented in algorithm notes (related TODO #108) |
| Breaking change | Default remains `List`; all existing tests must pass unchanged |
| `FastCollections` + `Array` combo | Array buffer unlocks existing fast paths automatically |
| Immutable/set types accidentally switched to array | Restrict `Array` to `TypeKind.Interface` sequence interfaces only |
